Spigot-API-Patches/0038-Add-proper-Entity-Player-OfflinePlayer-equals-and-ha.patch

```diff
@@ -0,0 +1,191 @@
++     * document the difference.
++     * @param other The specified object.
++     * @return {@code true} if this {@code Entity} is equal to the specified
++     *     object, {@code false} otherwise.
++     *
++     * @see Entity#getEntityId()
++     */
++    @Override
++    boolean equals(final Object other);
++
++    /**
++     * Returns the hash code for this {@code Entity}.
++     * The entity's hash code is defined to be as:
++     * <pre>
++     *     <code>    int hash = 7;
++     *         hash = (hash * 29) + getEntityId();
++     *     </code>
++     * </pre>
++     * Sub-interfaces are allowed to change these specifications,
++     * but they must document the difference.
++     * @return The hash code for this entity.
++     */
++    @Override
++    int hashCode();
++
++    // Paper end
++
+     // Spigot Start
+     public class Spigot
+     {
+diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
+index d636c63..e366e23 100644
+--- a/src/main/java/org/bukkit/entity/Player.java
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -1488,6 +1488,57 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
+     boolean hasResourcePack();
+     // Paper end
+ 
++    // Paper start
++    /*
++     * For backwards compatibility with old CraftPlayer.equals
++     */
++
++    /**
++     * Checks if this {@code Player} is equal to the specified object.
++     * <p>
++     *     This method will check for equality based upon the instance of the
++     *     specified object:
++     *     <ol>
++     *         <li>
++     *             If the instance is an instance of {@code Player}:
++     *             <ul>
++     *                 <li>
++     *                     If the entity ids and entity uuids are equal.
++     *                 </li>
++     *             </ul>
++     *         </li>
++     *         <li>
++     *             If the instance is an instance of {@code OfflinePlayer}
++     *             and not an instance of {@code Player}:
++     *             <ul>
++     *                 <li>
++     *                     If the uuids are equal, returned by
++     *                     {@link OfflinePlayer#getUniqueId()}.
++     *                 </li>
++     *             </ul>
++     *         </li>
++     *         <li>
++     *             Any other case will return {@code false}.
++     *         </li>
++     *     </ol>
++     * </p>
++     *
++     * Sub-interfaces are allowed to change these specifications,
++     * but they must document the difference.
++     *
++     * @param other The specified object.
++     * @return {@code true} if this {@code Player} is equal to the specified
++     *     object, {@code false} otherwise.
++     *
++     * @see Entity#getEntityId()
++     * @see Entity#getUniqueId()
++     * @see OfflinePlayer#getUniqueId()
++     */
++    @Override
++    boolean equals(final Object other);
++
```

Comment:
@kashike get carried away? This one looks pretty standard. @Searchndstroy this is fine.